idaes/apps/caprese/rolling.py

```diff
@@ -0,0 +1,267 @@
+                'tolerance.')
+        else:
+            # t_new > t_last + 2*tolerance
+            return tpoints
+
+    def extend(self, tpoints):
+        """
+        Extends by a valid iterable of time points.
+        """
+        tpoints = list(tpoints)
+        tpoints = self.validate_time(tpoints)
+        tpoints = self.validate_extend(tpoints)
+        super().extend(tpoints)
+
+    def find_nearest_index(self, target):
+        """
+        Returns the index of the nearest point in self.
+        """
+        # This is a copy of the same method of ContinuousSet.
+        # 
+        tol = self.tolerance
+        lo = 0
+        hi = len(self)
+        i = bisect.bisect_right(self, target, lo=lo, hi=hi)
+        # i is index at which target should be inserted if it is
+        # to be right of any equal components.
+
+        if i == lo:
+            # target is less than every entry of the set
+            nearest_index = i
+            delta = self[nearest_index] - target
+        elif i == hi:
+            nearest_index = i-1
+            delta = target - self[nearest_index]
+        else:
+            delta, nearest_index = min((abs(target - self[_]), _) 
+                    for _ in [i-1, i])
+
+        if delta > tol:
+            return None
+        return nearest_index
+
+
+    def binary_search(self, t):
+        """
+        """
+        # TODO: clean up this implementation
+        tolerance = self.tolerance
+
+        i = bisect.bisect_left(self, t)
+        t_ge = self[i]
+        if t_ge - t < tolerance:
+            return True, i
+
+        if i == 0:
+            return False, i
+
+        t_l = self[i-1]
+        if t - t_l < tolerance:
+            return True, i-1
+
+        return False, i
+
+
+class VectorSeries(OrderedDict):
+    """
+    A time-indexed vector.
+    """
+    # This class's purpose is suspiciously similar to that of IndexedComponent.
+    # Could have an OrderedDict of time-indexed variables/Parameters that I
+    # continually update along with the underlying time set.
+    def __init__(self, data=None, time=None, name=None, tolerance=0.0):
+        """
+        Args:
+            data: OrderedDict mapping cuids to lists of values
+            time: list of time points corresponding to data values
+            name: name of this vector
+        """
+        if data is None:
+            data = OrderedDict()
+        if time is None:
+            time = []
+
+        len_time = len(time)
+        for val_list in data.values():
+            if len(val_list) != len_time:
+                raise ValueError(
+                    'data lists and time list must have same length')
+
+        self.name = name
+        self.time = TimeList(time, tolerance=tolerance)
+
+        super().__init__(data)
+
+    def dim(self):
+        return len(self.keys())
+
+    def __len__(self):
+        # Behavior of len/dim could be confusing. This needs to be
+        # well-documented.
```

Comment:
Could you add docstrings to `__len__` and `dim` to address this comment?

Reply:
Done